src/lib/FlightTasks/tasks/Utility/ObstacleAvoidance.cpp

```diff
@@ -0,0 +1,227 @@
+		{0, {NAN, NAN, NAN}, {NAN, NAN, NAN}, {NAN, NAN, NAN}, NAN, NAN, false, {0, 0, 0}},
+		{0, {NAN, NAN, NAN}, {NAN, NAN, NAN}, {NAN, NAN, NAN}, NAN, NAN, false, {0, 0, 0}}
+	}
+};
+
+ObstacleAvoidance::ObstacleAvoidance(ModuleParams *parent) :
+	ModuleParams(parent)
+{
+	_desired_waypoint = empty_trajectory_waypoint;
+}
+
+ObstacleAvoidance::~ObstacleAvoidance()
+{
+	//unadvertise publishers
+	if (_pub_traj_wp_avoidance_desired != nullptr) {
+		orb_unadvertise(_pub_traj_wp_avoidance_desired);
+	}
+
+	if (_pub_pos_control_status != nullptr) {
+		orb_unadvertise(_pub_pos_control_status);
+	}
+
+}
+
+bool ObstacleAvoidance::initializeSubscriptions(SubscriptionArray &subscription_array)
+{
+	if (!subscription_array.get(ORB_ID(vehicle_trajectory_waypoint), _sub_vehicle_trajectory_waypoint)) {
+		return false;
+	}
+
+	if (!subscription_array.get(ORB_ID(vehicle_status), _sub_vehicle_status)) {
+		return false;
+	}
+
+	return true;
+}
+
+void ObstacleAvoidance::prepareAvoidanceSetpoints(Vector3f &pos_sp, Vector3f &vel_sp, float &yaw_sp,
+		float &yaw_speed_sp)
+{
+
+	if (!COM_OBS_AVOID.get() || _sub_vehicle_status->get().nav_state == vehicle_status_s::NAVIGATION_STATE_AUTO_LOITER) {
+		// if avoidance disabled or in failsafe nav_state LOITER, don't inject setpoints from avoidance system
+		return;
+	}
+
+	const bool avoidance_data_timeout = hrt_elapsed_time((hrt_abstime *)&_sub_vehicle_trajectory_waypoint->get().timestamp)
+					    > TRAJECTORY_STREAM_TIMEOUT_US;
+	const bool avoidance_point_valid =
+		_sub_vehicle_trajectory_waypoint->get().waypoints[vehicle_trajectory_waypoint_s::POINT_0].point_valid == true;
+	const bool is_loitering = _sub_vehicle_status->get().nav_state == vehicle_status_s::NAVIGATION_STATE_AUTO_LOITER;
+
+	if ((avoidance_data_timeout || !avoidance_point_valid) && !is_loitering) {
+		PX4_WARN("Obstacle Avoidance system failed, loitering");
+		_publish_vehicle_cmd_do_loiter();
+		return;
+	}
+
+	pos_sp = _sub_vehicle_trajectory_waypoint->get().waypoints[vehicle_trajectory_waypoint_s::POINT_0].position;
+	vel_sp = _sub_vehicle_trajectory_waypoint->get().waypoints[vehicle_trajectory_waypoint_s::POINT_0].velocity;
+	yaw_sp =  _sub_vehicle_trajectory_waypoint->get().waypoints[vehicle_trajectory_waypoint_s::POINT_0].yaw;
+	yaw_speed_sp = _sub_vehicle_trajectory_waypoint->get().waypoints[vehicle_trajectory_waypoint_s::POINT_0].yaw_speed;
+}
+
+void ObstacleAvoidance::updateAvoidanceDesiredWaypoints(const Vector3f &curr_wp, const float curr_yaw,
+		const float curr_yawspeed,
+		const Vector3f &next_wp, const float next_yaw, const float next_yawspeed)
+{
+	_desired_waypoint.timestamp = hrt_absolute_time();
+	_desired_waypoint.type = vehicle_trajectory_waypoint_s::MAV_TRAJECTORY_REPRESENTATION_WAYPOINTS;
+	_curr_wp = curr_wp;
+
+	curr_wp.copyTo(_desired_waypoint.waypoints[vehicle_trajectory_waypoint_s::POINT_1].position);
+	Vector3f(NAN, NAN, NAN).copyTo(_desired_waypoint.waypoints[vehicle_trajectory_waypoint_s::POINT_1].velocity);
```

Comment:
Recent addition: https://github.com/PX4/Matrix/pull/80/files#diff-b6575389f05f11bcd22c23d3e24bc7a6R415
```suggestion
	_desired_waypoint.waypoints[vehicle_trajectory_waypoint_s::POINT_1].velocity.setNaN();
```

Reply:
the matrix library submodule is stuck in January. I'll make a seperate PR to update the submodule and I will make the change

Reply:
#11706

Reply:
@MaEtUgR is `setNan()` supposed to work with type `float32[3]` ?

Reply:
Ah no sry, it's `Vector3f.setNan()`.
I was just trying to find out how to call https://github.com/PX4/Matrix/blob/716e96853f2c1f55c128799a380f3bdeb51662e0/matrix/Matrix.hpp#L520-L525 but you can also leave it unchanged.